src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    /// <summary>
+    /// DiagnosticHandler notifies DiagnosticSource subscribers about outgoing Http requests
+    /// </summary>
+    internal sealed class DiagnosticsHandler : DelegatingHandler
+    {
+        /// <summary>
+        /// DiagnosticHandler constructor
+        /// </summary>
+        /// <param name="innerHandler">Inner handler: Windows or Unix implementation of HttpMessageHandler. 
+        /// Note that DiagnosticHandler is the latest in the pipeline </param>
+        public DiagnosticsHandler(HttpMessageHandler innerHandler) : base(innerHandler)
+        {
+        }
+
+        protected internal override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,
+            CancellationToken cancellationToken)
+        {
+            //do not write to diagnostic source if request is invalid or cancelled,
+            //let inner handler decide what to do with it
+            if (request == null || cancellationToken.IsCancellationRequested)
+            {
+                return base.SendAsync(request, cancellationToken);
+            }
+
+            Guid loggingRequestId = LogHttpRequest(request);
+            Task<HttpResponseMessage> responseTask = base.SendAsync(request, cancellationToken);
+            LogHttpResponse(responseTask, loggingRequestId);
+            return responseTask;
+        }
+
+        #region private
+
+        private static readonly DiagnosticListener s_diagnosticListener =
+            new DiagnosticListener(DiagnosticsHandlerLoggingStrings.DiagnosticListenerName);
+
+        private static Guid LogHttpRequest(HttpRequestMessage request)
+        {
+            return s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.RequestWriteName) ?
+                LogHttpRequestCore(request) :
+                Guid.Empty;
+        }
+
+        private static void LogHttpResponse(Task<HttpResponseMessage> responseTask, Guid loggingRequestId)
+        {
+            if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ResponseWriteName))
+            {
+                ScheduleLogResponse(responseTask, loggingRequestId);
+            }
+        }
+
+        private static void ScheduleLogResponse(
+            Task<HttpResponseMessage> responseTask, Guid loggingRequestId)
+        {
+            responseTask.ContinueWith(
+                t =>
+                {
+                    if (!t.IsCanceled)
+                    {
+                        LogHttpResponseCore(t.IsFaulted ? null : t.Result, t.Exception, loggingRequestId);
+                    }
+                },
+                TaskScheduler.Default);
```

Comment:
A few things:
- Why are we not logging a response if the operation was canceled?  Seems like that'd still be interesting to know, as otherwise someone looking at a log doesn't know whether the operation is still in flight or whether it was canceled.
- If the intent really is though to not log anything if it's canceled, that's more efficiently achieved by passing TaskContinuationOptions.NotOnCanceled to the ContinueWith call.
- The lambda is also closing over loggingRequestId, which allocates both a closure and a delegate to a method on that closure object.  It'd be more efficient to pass the needed state in through the continuation's object state; since the state in this case is a Guid, it'll still be boxed, but that's better than the two larger allocations for the delegate and closure.  Something like:
```C#
responseTask.ContinueWith((t,s) =>LogHttpResponseCore(t.IsFaulted ? null : t.Result, t.Exception, (Guid)s),
    CancellationToken.None, TaskContinuationOptions.NotOnCanceled, TaskScheduler.Default);
```